### ISSUE TITLE
Add insertion iterator capabilities to DataSet

### ISSUE THREAD

**Issue by greenwald** [yap/YAP#61]
If we add insert iterators capabilities to DataSet, we can potentially fill data simply and easily. For example:

Say we have a function that returns a random point in the phase space of a model:
```cpp
template <class Generator>
std::vector<FourVector<double> > phsp(const Model& M, Generator& G);
```
and a `DataSet` `data`, `Model` `m`, and random number engine `g`, we could generate data with a simple statement:
```cpp
std::generate_n(std::back_inserter(data), n, std::bind(&phsp, m, g));
```

Likewise, given a function that returns a data point from an `istream_iterator`
```cpp
std::vector<FourVector<double> > next(std::istream_iterator is);
```
we could pull data from a file via something like
```cpp
std::generate(file.begin(), file.end(), std::back_inserter(data), std::function(next));
```

**Comment by greenwald** [comment]
So, after some investigation we know: One doesn't create his own `insert_iterator` or `back_insert_iterator`, one instead implements `insert` or `push_back` for the container and then uses the std iterators. So---

Given
```cpp
void DataSet::push_back(const std::vector<FourVector<double> >& P);
```

We can call

```cpp
std::back_inserter(dataSet);
```

So, if we want to have the iterators possible for our internal storage, which is a vector, we should implement:
```cpp
void DataSet::push_back(const std::vector<FourVector<double> >& P);
void DataSet::push_back(std::vector<FourVector<double> >&& P);
iterator DataSet::insert(iterator pos, const std::vector<FourVector<double> >& P);
iterator DataSet::insert(iterator pos, std::vector<FourVector<double> >&& P);
```

Two issues:
1. What is `iterator`? I think it should be `DataIterator`.
2. We still want to have one common function for creating a `DataPoint` _for_ the `DataSet`. So:
- Remove `DataSet` pointer from `DataPoint`
- Change `DataPoint::setFinalStateMomenta(...)` to be:
```cpp
void setFinalStateMomenta(const std::vector<FourVector<double> >& P, const Model& M, StatusManager& sm);
void setFinalStateMomenta(const std::vector<FourVector<double> >& P, DataSet& D);
```
The second function calls the first one, passing the `DataSet` as the `StatusManager` and passing `*D.model()` as the model, but first checks that `D.model()` is not nullptr.
- Add constructor to DataPoint that will create it and then fill it by calling `setFinalStateMomenta` on it.
- The above mentioned functions will call this constructor.

While you're at it, you might also think to implement:
* Move creation of empty point (currently inside `addEmptyPoint`) to function `createEmptyPoint`
* `addEmptyPoint` should now create empty point and add it

**Comment by greenwald** [comment]
You may also want to implement a `DataPoint` constructor that copies its structure from another `DataPoint` and then calls `setFinalStateMomenta`

**Comment by greenwald** [comment]
implemented in pull #66
